### PR TITLE
Us3 formular ingredients 

### PR DESCRIPTION
**Pull request by gwittm**

### CODE REVIEW

File: components/FormularAddRecipe.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            {
+              ingredientId: nanoid(),
+              name: event.target.elements.ingredient.value,
+              amount: event.target.elements.amount.value,
+              unit: event.target.elements.unit.value,
+            },
+          ]);
+        }}
+      >
+        {" "}
```
Comment (gregorsart) [inline components/FormularAddRecipe.js:148]: `Line 148` VSC makes this sometimes automatically, but it does not serve a purpose. (I would remove it)

File: components/FormularAddRecipe.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+`;
+
+const StyledButton = styled.button`
+  background-color: hotpink;
+  width: 480px;
+  margin: 10px;
+  height: 40px;
+  font-size: 18px;
+  border-radius: 5px;
+  align-item: center;
```
Comment (NikRoe) [inline components/FormularAddRecipe.js:50]:
```suggestion
  align-items: center;
```

File: components/FormularAddRecipe.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        </StyledInput>
+      </StyledForm>
+
+      <form
+        onSubmit={(event) => {
+          event.preventDefault();
+          setIngredients([
+            ...ingredients,
+            {
+              ingredientId: nanoid(),
```
Comment (NikRoe) [inline components/FormularAddRecipe.js:140]: your database will create an id automatically, there is no need for you to create another one manually

File: components/FormularAddRecipe.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
+import styled from "styled-components";
+import { useState } from "react";
+import { nanoid } from "nanoid";
```
Comment (NikRoe) [inline components/FormularAddRecipe.js:3]: you won't need this

File: components/FormularAddRecipe.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+          <input
+            defaultValue="1"
+            id="servings"
+            name="servings"
+            type="number"
+            min="1"
+          />
+        </StyledInput>
+        <StyledInput>
+          <label htmlFor="instruction">Instuction</label>
```
Comment (NikRoe) [inline components/FormularAddRecipe.js:123]:
```suggestion
          <label htmlFor="instruction">Instructions:</label>
```

File: components/FormularAddRecipe.js
Comment (NikRoe) [inline components/FormularAddRecipe.js:1]: This file has almost 200 lines of code, you should consider breaking your code up into smaller parts (Components).

Your form for the ingredients could be a separate component for instance.

Splitting your code up into smaller portions will make your project more maintainable.

File: components/FormularAddRecipe.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            id="title"
+            name="title"
+            type="text"
+            defaultValue={defaultData?.title}
+          />
+        </StyledInput>
+        <StyledInput>
+          <label htmlFor="preparation">
+            preparation:
+            <select name="preparation" defaultValue="Oven">
```
Comment (NikRoe) [inline components/FormularAddRecipe.js:81]: this select is missing its id-attribute to connect it with the label:

```suggestion
            <select name="preparation" defaultValue="Oven" id="preparation">
```

File: components/FormularAddRecipe.js [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+              <option value="Oven">Oven</option>
+              <option value="Microwave">Microwave</option>
+              <option value="Stove">Stove</option>
+              <option value="Grill">Grill</option>
+            </select>
+          </label>
+        </StyledInput>
+        <StyledInput>
+          <label htmlFor="course" defaultValue="Cake">
```
Comment (NikRoe) [inline components/FormularAddRecipe.js:90]:
```suggestion
          <label htmlFor="course">
```

File: components/FormularAddRecipe.js [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            Course:
+            <select name="course" defaultValue="Cake">
```
Comment (NikRoe) [inline components/FormularAddRecipe.js:92]:
```suggestion
            <select name="course" defaultValue="Cake" id="course">
```

File: components/FormularAddRecipe.js [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+              <option value="ml">ml</option>
+              <option value="tbsp">tbsp</option>
+              <option value="tsp">tsp</option>
+              <option value="pt">pt</option>
+              <option value="cup">cup</option>
+              <option value="pcs">pcs</option>
+              <option value="pn">pn</option>
+            </select>
+          </StyledIngredientInput>
+          <button>+</button>
```
Comment (NikRoe) [inline components/FormularAddRecipe.js:174]:
```suggestion
          <button type="submit">+</button>
```

File: db/models/Recipe.js [hunk 10]
```diff
@@ -3,13 +3,20 @@ import mongoose from "mongoose";
 const { Schema } = mongoose;
 
 const recipeSchema = new Schema({
-  title: { type: String, required: true },
+  /*   id: { type: String },
+   */ title: { type: String, required: true },
```
Comment (NikRoe) [inline db/models/Recipe.js:7]: you should remove all commented out code before creating a PR

```suggestion
title: { type: String, required: true },
```

File: db/models/Recipe.js [hunk 11]
```diff
@@ -3,13 +3,20 @@ import mongoose from "mongoose";
...
   preparation: { type: String },
-  // image: { type: String },
   course: { type: String },
   time: { type: String },
   servings: { type: Number },
-  // ingredients: { type: Array },
+  ingredients: [
+    {
+      ingredientId: { type: String },
```
Comment (NikRoe) [inline db/models/Recipe.js:14]: this will be created automatically by your database

File: package.json [hunk 12]
```diff
@@ -18,7 +18,9 @@
     "react": "18.2.0",
     "react-dom": "18.2.0",
     "styled-components": "^6.0.2",
-    "swr": "^2.2.4"
+    "swr": "^2.2.4",
+    "uid": "^2.0.2",
+    "uuid": "^9.0.1"
```
Comment (NikRoe) [inline package.json:23]: do you actually need uid or uuid? If not, remove both libraries

File: pages/index.js [hunk 13]
```diff
@@ -7,39 +7,56 @@ const List = styled.ul`
...
-              <br></br>
-              {recipe.time}
-              <br></br>
-              {recipe.servings}
-              <br></br>
-              {recipe.instruction} */}
+              {/*  <p>Ingredients:</p>
+              {recipe.ingredients.map((ingredient, index) => (
+                <div key={index}>{ingredient.name}</div>
+              ))} */}
```
Comment (NikRoe) [inline pages/index.js:53]: you should remove all commented out code before creating a PR

```suggestion
```

File: pages/index.js [hunk 14]
```diff
@@ -7,39 +7,56 @@ const List = styled.ul`
...
   const { data } = useSWR("/api/recipes", { fallbackData: [] });
+
   return (
-    <>
-      <h1>Rezepte</h1>
+    <StyledBox>
       <List role="list">
-        {data.map((recipe) => {
+        <h1>Rezepte</h1>
+        {data.map((recipe, _id) => {
```
Comment (NikRoe) [inline pages/index.js:46]: what purpose does the `_id` serve here? I think you can remove it

```suggestion
        {data.map((recipe) => {
```

**Review by NikRoe — CHANGES_REQUESTED** [review]
Hi there,

this looks like a very promising start 👏 

I left a few comments in places where there is still some room for improvement.

Keep it up 🙌

File: components/FormularAddRecipe.js (hunk 3 above)
Comment (gwittm) [reply components/FormularAddRecipe.js:140]: At this point we do not send the ingredients to our database. Without adding an id we got the note "each child in a list needs an unique key-prop". Later we want to be able the delete items within this formular.

Comment (NikRoe) [reply components/FormularAddRecipe.js:140]: You could work with the name of the ingredient for the key prop.

You should only work on what is relevant for the current user story and not take features of an upcoming user story into consideration.

**Comment by gwittm** [comment]
thank you for your help. We changed everything accordingly :)

File: components/Formular/FormularAddRecipe.js [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+            return (
+              <li key={ingredient.name}>
+                {ingredient.name} {""} {ingredient.amount}
+                {ingredient.unit}
+              </li>
+            );
+          })}
+        </ul>
+      </StyledIngredientsSection>
+      <StyledButton type="submit" form="recipeForm" onSubmit={handleSubmit}>
```
Comment (NikRoe) [inline components/Formular/FormularAddRecipe.js:110]: you don't need the handleSubmit here:

```suggestion
      <StyledButton type="submit" form="recipeForm">
```

File: components/Formular/FormularIngredients.js [hunk 16]
```diff
@@ -0,0 +1,52 @@
+import {
+  StyledIngredientsSection,
+  StyledIngredientInput,
+} from "./FormularStyling.js";
+import { useState } from "react";
```
Comment (NikRoe) [inline components/Formular/FormularIngredients.js:5]: you can remove this import

File: components/Formular/FormularIngredients.js [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+import { nanoid } from "nanoid";
+
+export default function FormularIngredients({ onAddIngredient }) {
+  return (
+    <>
```
Comment (NikRoe) [inline components/Formular/FormularIngredients.js:10]: you don't need those fragments anymore

**Review by NikRoe — CHANGES_REQUESTED** [review]
Hi there,

nicely done and almost there 🙌 

I left some comments that will help tidy up a little =)

File: components/Formular/FormularAddRecipe.js [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+          ></textarea>
+        </StyledInput>
+      </StyledForm>
+      <FormularIngredients onAddIngredient={handleAddIngredient} />
+      <StyledIngredientsSection>
+        <p>added Ingredients:</p>
+        <ul>
+          {ingredients.map((ingredient) => {
+            return (
+              <li key={ingredient.IngredientId}>
```
Comment (NikRoe) [inline components/Formular/FormularAddRecipe.js:102]: In your Schema this key is written differently:

```suggestion
              <li key={ingredient.ingredientID}>
```

**Review by NikRoe — APPROVED** [review]
Hi there,

there a minor issue with your key for the ingredients. It needs to match what you define in your Schema.

Since it's only a very small thing, go ahead and change it and process to the QA without having to repost this PR

Best regards 👋

**Pull request merged**